### PR TITLE
SD polyfill docs

### LINKED ISSUE #9761: Amp shadow - ShadowDOM v1 slots not supported
ShadowDom V1 slots are not supported by firefox or safari < 10. And don't seem to be working with the polyfills that come with AmpShadow

I would usually polyfill these, but it seems that the shadyDom and ShadowDom polyfill are incompatible with AmpShadow.

### PR DESCRIPTION
**Pull request by dvoytenko**
Closes #9761.

/cc @rowangray

### CODE REVIEW

File: spec/amp-shadow-doc.md [hunk 1]
```diff
@@ -96,6 +96,42 @@ Both `AMP.attachShadowDoc` and `AMP.attachShadowDocAsStream` return a `ShadowDoc
 - `shadowDoc.close()` - closes the AMP document and frees the resources.
 
 
+## Shadow DOM API and polyfills
+
+AMP Shadow Docs rely heavily on [Shadow DOM API](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Shadow_DOM). This is a powerful and elegant API, part of Web Components family. It allows natural isolation between major parts of the page and, as such, an ideal tool for PWAs and AMP Shadow Docs.
```
Comment (pbakaus) [inline spec/amp-shadow-doc.md]: on -> on the

Comment (pbakaus) [reply spec/amp-shadow-doc.md]: part of -> part of the

Comment (pbakaus) [reply spec/amp-shadow-doc.md]: It allows -> it allows for

Comment (pbakaus) [reply spec/amp-shadow-doc.md]: an ideal -> is an ideal

File: spec/amp-shadow-doc.md [hunk 2]
```diff
@@ -96,6 +96,42 @@ Both `AMP.attachShadowDoc` and `AMP.attachShadowDocAsStream` return a `ShadowDoc
...
+
+Shadow DOM is currently implemented in Chrome and newer Safari, so it has a solid modern support and definitely not ubiquitous. AMP Shadow Docs API polyfills the necessary parts of Shadow DOM.
```
Comment (pbakaus) [inline spec/amp-shadow-doc.md]: "solid modern support and definitely not ubiquitous"...so which is it? :) maybe reword to something like "Shadow DOM is currently only implemented in Chrome and newer versions of Safari, so in order to support all of your users, the AMP Shadow Docs API relies on polyfills for necessary parts of Shadow DOM."

File: spec/amp-shadow-doc.md [hunk 3]
```diff
@@ -96,6 +96,42 @@ Both `AMP.attachShadowDoc` and `AMP.attachShadowDocAsStream` return a `ShadowDoc
...
+
+However, not all advanced Shadow DOM features are polyfilled by AMP Shadow Docs. In particular, [shadow slots](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Slot) are not polyfilled. If you'd like to use slots and similar advanced features, please use one of the Shadow DOM polyfill, such as [WebComponents.js](https://github.com/webcomponents/webcomponentsjs). If you do, we recommend the following code structure (using Web Components polyfill as an example):
```
Comment (pbakaus) [inline spec/amp-shadow-doc.md]: polyfill -> polyfills

File: spec/amp-shadow-doc.md [hunk 4]
```diff
@@ -96,6 +96,42 @@ Both `AMP.attachShadowDoc` and `AMP.attachShadowDocAsStream` return a `ShadowDoc
...
+
+HTML:
+```html
+<script async src="https://cdn.ampproject.org/shadow-v0.js"></script>
+<script async src="https://cdnjs.cloudflare.com/ajax/libs/webcomponentsjs/1.0.3/webcomponents-sd-ce.js"></script>
```
Comment (pbakaus) [inline spec/amp-shadow-doc.md]: How does Shadow AMP ensure that the web components polyfill is used? How does it detect it? Couldn't it load before the polyfill here?

Comment (torch2424) [reply spec/amp-shadow-doc.md]: I helped build out the AMP Shadow with Polyfill solution. Currently, we would tap into the `'WebComponentsReady'` event that is broadcasted to the window.

If you are speaking of the use case where one would not like not use the webcomponents polyfill, but would still like to use the amp ShadowDOM polyfill, then they would simpley not include the `sdReadyPromise` snippet of code.

Is this what you meant?

File: spec/amp-shadow-doc.md (hunk 1 above)
Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: Done

File: spec/amp-shadow-doc.md (hunk 2 above)
Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: Done

File: spec/amp-shadow-doc.md (hunk 3 above)
Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: Done

File: spec/amp-shadow-doc.md (hunk 4 above)
Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: It would work mostly correctly if AMP APIs are used before polyfill is loaded. However, below I recommend waiting until polyfill is loaded. See the part with `const sdReadyPromise =`. It's the most reliable way to ensure that (a) things are loaded as much in parallel as possible, and (b) the app waits for right moment.

Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: @torch2424 The idea is that if another polyfill is used, the `sdReadyPromise` would still be used, but it'd instead be switched according to the other polyfill API.

File: spec/amp-shadow-doc.md (hunk 3 above)
Comment (torch2424) [inline spec/amp-shadow-doc.md]: I remember that we were worried that the developers may try to use their own or another less reputable polyfill for Shadow DOM slots, and it may conflict with our current implementation. 

I think it would be good to mention that the webcomponents script was the polyfill that we tested for `pwa.js` and we recommend as we can confirm that it works. I feel like the wording suggests that any polyfill will work (which it definitely should), but it would be better for someone trying AMP Shadow DOM for the first time.

File: spec/amp-shadow-doc.md (hunk 4 above)
Comment (torch2424) [reply spec/amp-shadow-doc.md]: Ah yes that makes sense, thanks :)

File: spec/amp-shadow-doc.md (hunk 3 above)
Comment (dvoytenko) [reply spec/amp-shadow-doc.md]: I made notes that we tested with Web Components polyfill and invited to report other polyfills. We'll consider including docs for others depending on popularity and quality.

**Comment by dvoytenko** [comment]
@pbakaus @torch2424 Updates have been made. PTAL.

**Comment by torch2424** [comment]
Thanks for writing these dos :)
